crates/storage/src/traits/storage.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2018-2022 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use ink_primitives::Key;
+
+/// Types that implement `scale::Encode` and `scale::Decode` called - **Packed**. Those types
+/// support serialization and deserialization into/from storage and occupy only one storage cell.
+///
+/// All other types - **Non-Packed**.
+///
+/// # Note
+///
+/// The trait is automatically implemented for types that implement `scale::Encode`
+/// and `scale::Decode` via blank implementation.
+///
+/// Don't try to implement that trait manually.
+pub trait Packed: scale::Decode + scale::Encode {}
+
+/// Every type that wants to be a part of the storage should implement this trait.
+/// The trait is used for serialization/deserialization into/from storage.
+///
+/// # Note
+///
+/// The trait is automatically implemented for [`Packed`](crate::traits::Packed) types
+/// via blank implementation.
+pub trait Storable: Sized {
+    /// Convert self to a slice and append it to the destination.
+    fn encode<T: scale::Output + ?Sized>(&self, dest: &mut T);
+
+    /// Attempt to deserialize the value from input.
+    fn decode<I: scale::Input>(input: &mut I) -> Result<Self, scale::Error>;
+}
+
+/// Returns storage key for the type
+///
+/// # Note
+///
+/// The trait is automatically implemented for [`Packed`](crate::traits::Packed) types
+/// via blank implementation.
+pub trait KeyHolder {
+    /// Storage key of the type
+    const KEY: Key;
+
+    /// Returns the storage key.
+    fn key(&self) -> Key {
+        Self::KEY
+    }
+}
+
+/// Describes the type that should be used for storing the value and preferred storage key.
+///
+/// # Note
+///
+/// The trait is automatically implemented for [`Packed`](crate::traits::Packed) types
+/// via blank implementation.
+pub trait Item<Salt: KeyHolder> {
+    /// Type with storage key inside
+    type Type: Storable;
+    /// Preferred storage key
+    type PreferredKey: KeyHolder;
+}
+
+/// Automatically returns the type that should be used for storing the value.
+///
+/// Trait is used be codegen to use the right storage type.
```

Comment:
"... used _by_ codegen ..."

What does "correct" mean here?

Reply:
I think we should find a better name for `AutoItem`, something that reflects more what it's about. Maybe `StorageType`?

Regarding
>The trait is used by codegen to determine which storage key the type should have.

The word `type` here is confusing to me. We don't derive the key from the type, no? If I look at `fn convert_into_storage_field` the type is not part of the `key`.

Reply:
> The word type here is confusing to me. We don't derive the key from the type, no? If I look at fn convert_into_storage_field the type is not part of the key.


The type defines which storage key we should use=)

```rust
// Here, the storage key is `123`
let storage: Contract<ManualKey<123>> = Default::default(); 
// Here, the storage key is `234`
let storage: Contract<ManualKey<234>> = Default::default();
// Here, the storage key is a hash of `234` and `123`
let storage: Contract<ManualKey<234, ManualKey<123> /* It is Salt or ParentStorageKey */ >> = Default::default();
```

> I think we should find a better name for AutoItem, something that reflects more what it's about. Maybe StorageType?

`StorageType` is the previous name of the `Item`=) And previously, it was `AutoStorageType`. I'm okay with `StorageType` but it doesn't highlight that it is related to automatically type determination. `Item` helps to determine the type too, but it also contains `PreferredKey`

Reply:
What do you think about?:

```rust
pub trait Item<AutoGenerated: StorageKey> {
    type Type: Storable;
    type PreferredKey: StorageKey;
}

pub trait StorageType<AutoGenerated: StorageKey> {
    type Type: Storable;
}
```

or maybe simplify to `Generated`